src/api.h

```diff
@@ -39,25 +34,325 @@ AWS_CRT_API const char *aws_crt_error_debug_str(int);
 
 /* IO */
-AWS_CRT_API struct aws_event_loop_group *
-aws_crt_event_loop_group_new(uint16_t max_threads);
-AWS_CRT_API void
-aws_crt_event_loop_group_release(struct aws_event_loop_group *elg);
+typedef struct aws_event_loop_group aws_crt_event_loop_group;
+typedef struct _aws_crt_event_loop_group_options aws_crt_event_loop_group_options;
+AWS_CRT_API aws_crt_event_loop_group_options *aws_crt_event_loop_group_options_new(void);
+AWS_CRT_API void aws_crt_event_loop_group_options_release(aws_crt_event_loop_group_options *options);
+AWS_CRT_API void aws_crt_event_loop_group_options_set_max_threads(
+    aws_crt_event_loop_group_options *options,
+    uint16_t max_threads);
+AWS_CRT_API aws_crt_event_loop_group *aws_crt_event_loop_group_new(aws_crt_event_loop_group_options *options);
+AWS_CRT_API void aws_crt_event_loop_group_release(aws_crt_event_loop_group *elg);
+
+/* Input stream */
+typedef struct aws_input_stream aws_crt_input_stream;
+typedef struct _aws_crt_input_stream_options aws_crt_input_stream_options;
+typedef struct aws_stream_status aws_crt_stream_status;
+typedef int(aws_crt_input_stream_seek_fn)(void *user_data, int64_t offset, int seek_basis);
+typedef int(aws_crt_input_stream_read_fn)(void *user_data, uint8_t *dest, size_t dest_length);
+typedef int(aws_crt_input_stream_get_status_fn)(void *user_data, aws_crt_stream_status *status);
+typedef int(aws_crt_input_stream_get_length_fn)(void *user_data, int64_t *out_length);
+typedef int(aws_crt_input_stream_destroy_fn)(void *user_data);
+AWS_CRT_API aws_crt_input_stream_options *aws_crt_input_stream_options_new(void);
+AWS_CRT_API void aws_crt_input_stream_options_release(aws_crt_input_stream_options *options);
+AWS_CRT_API void aws_crt_input_stream_options_set_user_data(aws_crt_input_stream_options *options, void *user_data);
+AWS_CRT_API void aws_crt_input_stream_options_set_seek(
+    aws_crt_input_stream_options *options,
+    aws_crt_input_stream_seek_fn *seek_fn);
+AWS_CRT_API void aws_crt_input_stream_options_set_read(
+    aws_crt_input_stream_options *options,
+    aws_crt_input_stream_read_fn *read_fn);
+AWS_CRT_API void aws_crt_input_stream_options_set_get_status(
+    aws_crt_input_stream_options *options,
+    aws_crt_input_stream_get_status_fn *get_status_fn);
+AWS_CRT_API void aws_crt_input_stream_options_set_get_length(
+    aws_crt_input_stream_options *options,
+    aws_crt_input_stream_get_length_fn *get_length_fn);
+AWS_CRT_API void aws_crt_input_stream_options_set_destroy(
+    aws_crt_input_stream_options *options,
+    aws_crt_input_stream_destroy_fn *destroy_fn);
+
+AWS_CRT_API aws_crt_input_stream *aws_crt_input_stream_new(aws_crt_input_stream_options *options);
+AWS_CRT_API void aws_crt_input_stream_release(aws_crt_input_stream *input_stream);
+
+/* HTTP */
+typedef struct _aws_crt_http_headers aws_crt_http_headers;
+AWS_CRT_API aws_crt_http_headers *aws_crt_http_headers_new_from_blob(uint8_t *blob, size_t blob_length);
```

Comment:
A bit confused by your approach to byte-cursors in this API. which of these is your approach:

A) use them everywhere possible (if so use one here)
B) do not use them (if so, you're still using them in some places)
C) use them as a way to return (ptr,len) pairs because you can't return two values, but not as input because it's not as natural when writing bindings

Reply:
also, trivial: `const uint8_t *`